src/serializer/functions.js

```diff
@@ -140,16 +127,38 @@ export class Functions {
         let value = property.descriptor && property.descriptor.value;
         invariant(value !== undefined);
         invariant(value instanceof Value);
-        let entry = this.__optimizedFunctionEntryOfValue(value);
+        let entry = this._optimizedFunctionEntryOfValue(value);
         if (entry) recordedAdditionalFunctions.push(entry);
       }
     }
     return recordedAdditionalFunctions;
   }
 
+  _generateOptimizedFunctionsFromRealm(): Array<AdditionalFunctionEntry> {
+    let realm = this.realm;
+    let recordedAdditionalFunctions = [];
+    for (let [valueToOptimize, argModel] of realm.optimizedFunctions) {
+      let value = valueToOptimize instanceof AbstractValue ? this._unwrapAbstract(valueToOptimize) : valueToOptimize;
+      // Check for case where __optimize was called in speculative context where effects were discarded
+      if (value._isPartial === undefined) {
```

Comment:
Instead of obscure `_isPartial` check, introduce an `isValid()` like here: #2278

Is there a test for this?

Reply:
I don't think I can come up with a stable unit test case for this because iirc the only time that we throw away effects is when we end a speculative execution with an `InfeasiblePathError` which is dependent on the simplifier.